### DISCUSSION TITLE
Help with playback files download

### DISCUSSION THREAD

**Discussion by DenisMartin0** [thirtythreeforty/neolink#149] [Q&A]
Hi,

I know that it's not the main purpose of neolink, but is anyone able to decode the xml parameters to download "PlaybackFiles" request/response ?
Based on provided Wireshark plugin, it seems that the Windows client use type 13 and then type 8 to download a Playback file.
Unfortunately the plugin didn't decode the parameters, or I'm not doing it the right way... :

![image](https://user-images.githubusercontent.com/36104617/114462479-63d95100-9be3-11eb-9b41-1f8c0b9aed7f.png)

NB: I'm connected to a NVR, not a camera.

**Comment by QuantumEntangledAndy** [comment]
I believe the binary data is in the same format as the streams that we decode. It is discussed here in the project `dissector/mediapacket.md`. I think there are a few programs that can decide the data once saved to disk but I've never seen it decoded from the packets.

If your looking to write a program or extend neolink to decode them then you can find the bulk of the decoding here `src/bc_protocol/media_packet.rs` and here `src/bc_protocol.rs`.

**Reply by DenisMartin0** [reply]
You're probably talking about the PlaybackFile stream itself. This will be my next step :)
Currently I'd like to decode first the PlaybackFiles list, and then the download request for a specific file.
After I'll switch to playback-file decode.

I'll continue to play with Wireshark & Lua for a while.

**Reply by QuantumEntangledAndy** [reply]
Oh sorry the list. That's what I get for multitasking. I've never actually looked into the details of this packet as I've never plugged a sdcard in. If your willing to share the wireshark packets I could have a little look at them if time permits.

**Reply by DenisMartin0** [reply]
Let's take a simplier one that you've already seen: Compression Request/Response. It's documented in message.md, message #56
It seems that my wireshark setup is not able to decode it... :
![image](https://user-images.githubusercontent.com/36104617/114546042-01bd3200-9c5d-11eb-913a-ee0c3b16d339.png)

Here is the complete request packet:
ec71db54ba20049226dade070800450000a8808d400080060000c0a80081c0a800e7c6772328d7c8a50fcc4f63d85018201483530000f0debc0a3800000068000000000000150000146468000000d9216409ed100429bdf4a59e2c395dbf31aa494e69e07acf6c6c582cb28e6ad56a399a345a3a1443acc7191925304f2f05dd55e273a0da9b707af05f5d9eeac8c9368df7fb204d8c66cb44bb0903a5bca8eb1277c72fe02f0385553f6119c36e1688f51ffd6ad87c

Am I doing it something wrong ? Or maybe the "encryption" changed with latest firmware...

Note: I'm almost certain it's xml behind 'cause length of the message match what's expected... but i'm having no success with the existing decode function.

**Reply by DenisMartin0** [reply]
Additionally here is what's expected:
```
<?xml version="1.0" encoding="UTF-8" ?>
<Extension version="1.1">
<channelId>0</channelId>
</Extension>
```

Here is hex dump: 
```
d9216409ed100429bdf4a59e2c395dbf31aa494e69e07acf6c6c582cb28e6ad56a399a345a3a14 43
acc7191925304f2f05dd55e273a0da9b707af05f5d9eeac8c9 36
8df7fb204d8c66cb44bb0903a5bca8eb1277c72fe02f0385 55
3f6119c36e1688f51ffd6ad8 7c
```

**Reply by QuantumEntangledAndy** [reply]
Make sure your using the latest lua plugin. Decryption works by looking for the `<?xml` tag. Regardless of data it tries to decode the first few bytes and if it sees that tag it will decode the rest. But maybe something unexpected is happening with your data to stop that.

**Reply by DenisMartin0** [reply]
I'm using this one:
![image](https://user-images.githubusercontent.com/36104617/114554023-90827c80-9c66-11eb-87f8-a16499f2775d.png)

When I manually try to decode the first bytes of the upper buffer (D9 21 64...), it's not looking like "<?xml"

The firmware I'm using is from Feb 2021... Any chance that the 'crypto' changed recently ?
![image](https://user-images.githubusercontent.com/36104617/114554597-32a26480-9c67-11eb-94a0-361be8762a02.png)

**Reply by QuantumEntangledAndy** [reply]
Perhaps but we should rule out other things before looking for new encryptions. I belive the encryption key index is dependent on channel ID. Could you put the camera on channel 0 and see if that helps?

**Reply by QuantumEntangledAndy** [reply]
Sorry I just saw you were on channel zero for some reason I thought you were on channel 1. Nevermind.

**Reply by QuantumEntangledAndy** [reply]
Hmm interesting. Perhaps a few more details will help. 
1. Whats the camera we are connecting to (on the other end of the NVR).
2. Do any packets decode correctly?

**Reply by DenisMartin0** [reply]
No packets are decoded correctly.
Camera is a 12MP POE : RLC-1220A

ok, I've restarted the Windows client, and this is what I received respectively for Channel 0 and Channel 1:

![image](https://user-images.githubusercontent.com/36104617/114564667-d04e6180-9c70-11eb-9f23-36d2f57feb17.png)
`002c3bbaadd1b8cf49dddadc6a4d8537eda7621b4a3c8412e5555fc53df967b246f97ffa0fd3a0b19cba637cf8b2d4e77b85f14f7524ff6efb96fdd86364766c980b253046676c782bc5f76b3659bab4d93981fa0bec9d77e110b54ec8b7485652c54a1e1fb65935`

![image](https://user-images.githubusercontent.com/36104617/114564701-dcd2ba00-9c70-11eb-8947-c0cb8758d773.png)
`002c3bbaadd1b8cf49dddadc6a4d8537eda7621b4a3c8412e5555fc53df967b246f97ffa0fd3a0b19cba637cf8b2d4e77b85f14f7524ff6efb96fdd86364766c980b253046676c782bc5f76b3658bab42e5481e5d570002b34d742283e280d12fe47517eca172fa7`

Notes:
- The two messages are similar from the begining until a point (77 bytes), which is the position of actual ChannelId.
- They are both different from last time :)

**Reply by QuantumEntangledAndy** [reply]
3. What's the client you are capturing packets for.

**Reply by DenisMartin0** [reply]
Windows client:
![image](https://user-images.githubusercontent.com/36104617/114570142-e14da180-9c75-11eb-8898-bdbca42128d3.png)

**Reply by QuantumEntangledAndy** [reply]
Can you connect with neolink? I wonder if there is a change in firmware and if it selects the old encoder if requests are made with the old encoder (such as with neolink)

**Reply by QuantumEntangledAndy** [reply]
@thirtythreeforty any ideas what might be happening here. I can confirm it is not decoding with my dissector either. Not even if I force it to decode.

**Reply by DenisMartin0** [reply]
1/ The new encoding may vary with every TCP connection.
At init, the Win client create two different connections with the NVR (don't know why), and one of them is closed after a few secs.

Both go through Login, then VersionInfo then a pack with SysGeneral+AbilityInfo+AbilitySupport,...
The first encoded frame is AbilityInfo, and it's encoded a different way for the two connections.

2/ The new encoding re-use byte N-1 to encode byte N, as two XML msgs that differs only on one single byte (ChannelID) are encoded differently since this point up to the end. To my understanding this wasn't the case with currently documented protocol

**Reply by DenisMartin0** [reply]
I downloaded an older WinClient version from [here](https://support.reolink.com/hc/en-us/articles/360012466173-All-Reolink-Client-Versions) and it's decoding well :)

**Reply by thirtythreeforty** [reply]
Nice find!

The new firmware builds that Reolink are distributing have a suspicious-sounding "Optimized network transmission protocol to improve network security" changelog note.  With the roll-forward of firmware _and_ client versions, I'd say it's possible they're opting into a new crypto/obfuscation protocol.  Especially since the older client continues to decode.

Two questions:

- Is your camera running newer firmware? Could you repeat the exercise direct-connecting to the camera?
- Since the Login and VersionInfo request/response decode correctly, that means that if the new-crypto hypothesis is correct, they would have to negotiate this in one of those two messages. Can you post the difference of these messages between client revisions?

**Reply by DenisMartin0** [reply]
Very first request (ie: Login) is different:
```
Old client: f0debc0a010000002c0700000000000101dc1465
New client: f0debc0a010000002c0700000000000103dc1465
```
as you can notice, there is a 03 instead of a 01 in the Encrypt field:
![image](https://user-images.githubusercontent.com/36104617/114586379-e1a16900-9c84-11eb-995d-e2783d577fa7.png)

**Reply by thirtythreeforty** [reply]
Perfect! $5 says this is a new crypto scheme. I'll diff their new firmware in Ghidra and see if they've just changed the Charlie key.  Possibly it's now something better, but with the requests so similar, they don't seem to be negotiating RSA or TLS.

**Reply by DenisMartin0** [reply]
I'm expecting a little bit more change than only the key (but nothing close to RSA or TLS ;) ):
- Encoding changes at every new connection and it was not the case with 'old' version (Maybe based on PortNumber, or Timestamp,...)
- When changing a single byte in a message, it affect every bytes after this one and it was not the case before (Some kind of recursivity)

**Reply by thirtythreeforty** [reply]
This may also be a good opportunity to rework the existing decoders into Tokio message streams.  I'm not really happy with the legibility of my Baichuan parser.

**Reply by DenisMartin0** [reply]
By patching the `libBCSDKWrapper.dll` of the **new** client, I managed to force it to talk the **old** protocol.
Search for 0x03dc1465 occurrence in the file, there is only one.
Offset to patch: 0x0FF30B, Replace 0x03 by 0x01 and your done.

**Reply by QuantumEntangledAndy** [reply]
If you do rework the tokio parsers now would be a good time to include the message status codes and message IDs.

Also we had an issue recently where someone seemed to hit a limit on the number of cameras a single neolink can handle. They solved it by running multiple neolinks. I am wondering if we can start the TCP connections and parse them on different threads based on which camera we are talking too in order to alleviate this.

**Reply by QuantumEntangledAndy** [reply]
I have a branch that includes status codes and message ids as well as handeling the encryption flag in the login message correctly. Currently the branch is either yes or no encryption based on the login fields encryption byte but it is a quick rework to make it match based on the byte instead the three diffent values. 0=>no encryption, 1=>old enc, 3=>new enc

**Reply by thirtythreeforty** [reply]
@DenisMartin0 would you be willing to send me (thirtythreeforty@gmail.com) those two packet captures, new-encryption and old-encryption? I will keep them confidential.

Based on casual reverse engineering of the new firmware, I suspect they've started using AES. This shouldn't be too bad to implement if so, mainly a matter of tracking down either a) the hardcoded key or b) the key exchange (possibly obfuscated) in the login reply. @DenisMartin0 since you patched the client (ha! great job!) and the camera replied with old crypto, this means that it's the camera's job to pick the crypto parameters and the client simply follows suit.

(AES also lines up with your comment about the entire message shifting... a change in an AES message will affect the whole message.)

**Reply by thirtythreeforty** [reply]
I will also say... the way they're using AES is not secure either, for a couple reasons. So this isn't exactly a security upgrade.

**Reply by QuantumEntangledAndy** [reply]
@twistedddx sent me an email a while ago where he was looking through the firmware update and found this. In a part related to possible future AES encryption:

```
write   !shenzhenbaichuan.com@20161111!&    open tgz file failed.   open    /
mnt/tmp/config.aes open aes file failed.   BCSC    write aes file failed.  Unable to set encryption key in AES...  ppram == NULL   get_new_secretcode fail secret_code set fail    get config failed.  invalid chn_num(%d)
 
 
The !shenzhenbaichuan.com@20161111!&    part may be useful.
```

**Reply by DenisMartin0** [reply]
I'm playing with Ghidra and the dll. I'm almost certain it's AES128 from OpenSSL, probably CFB (to be confirmed).
I know the InitialVector, still looking for the Key.
![image](https://user-images.githubusercontent.com/36104617/114771354-bc882580-9d6c-11eb-872e-14ec9bdb72ef.png)

The Key might be in the nonce from the first login (This nonce seems different from the older one).

@thirtythreeforty I'll try to recapture two packets for you.

**Reply by QuantumEntangledAndy** [reply]
So I updated my firmware and I am getting the fun new packets too. Heres a wireshark dump of the initial packets. I've already changed all passwords so this should be safe to upload.


[EncryptionProtocol02_login.pcapng.zip](https://github.com/thirtythreeforty/neolink/files/6314731/EncryptionProtocol02_login.pcapng.zip)

**Reply by QuantumEntangledAndy** [reply]
First message send is the same expect that the staus code is `03dc` rather than `01dc`

|    magic     |  message id  | message length | encryption offset | encrypt | unknown | message class |
|--------------|--------------|----------------|-------------------|---------|---------|---------------|
| f0 de bc 0a  | 01 00 00 00  |  2c 07 00 00   |    00 00 00 01    |    03   |    dc   |     14 65     |


Camera then replies with `02dd` rather than `01dd`
|    magic     |  message id  | message length | encryption offset | encrypt | unknown | message class |
|--------------|--------------|----------------|-------------------|---------|---------|---------------|
| f0 de bc 0a  | 01 00 00 00  |  91 00 00 00   |    00 00 00 01    |    02   |    dd   |     14 66     |

Xml attached to this is still decryptable with the old scheme with the nonce being `AB5B32EC757765C2` in my case

No other changes in first message

**Reply by QuantumEntangledAndy** [reply]
The next message (the ModernLogin) is the same in the header

|    Magic     |  Message ID  | Message Length | Encryption Offset |    Status Code    | Message Class | Binary Offset |
|--------------|--------------|----------------|-------------------|-------------------|---------------|---------------|
| f0 de bc 0a  | 01 00 00 00  |  28 01 00 00   |    00 00 00 01    |       00 00       |     14 64     |  00 00 00 00  |

And the same in the xml payload which can still be decrypted fine

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<body>
<LoginUser version="1.1">
<userName>0DFB4D21F2A282128100524E2C4792B</userName>
<password>EC1AF34630047FEB2C5B9D2C5402A8D</password>
<userVer>1</userVer>
</LoginUser>
<LoginNet version="1.1">
<type>LAN</type>
<udpPort>0</udpPort>
</LoginNet>
</body>
```

The reply from the camera is the same in the header except the length of the message which is longer

|    Magic     |  Message ID  | Message Length | Encryption Offset |    Status Code    | Message Class | Binary Offset |
|--------------|--------------|----------------|-------------------|-------------------|---------------|---------------|
| f0 de bc 0a  | 01 00 00 00  |  2e 06 00 00   |    00 00 00 01    |       c8 00       |     00 00     |  00 00 00 00  |

The xml payload here is also perfectly decryptable using the current scheme but there are a few interesting extra xml keys

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<body>
<DeviceInfo version="1.1">
<firmVersion>00000000065536</firmVersion>
<IOInputPortNum>0</IOInputPortNum>
<IOOutputPortNum>0</IOOutputPortNum>
<diskNum>0</diskNum>
<type>wifi_solo_ipc</type>
<channelNum>1</channelNum>
<audioNum>1</audioNum>
<ipChannel>0</ipChannel>
<analogChnNum>1</analogChnNum>
<resolution>
<resolutionName>2304*1296</resolutionName>
<width>2304</width>
<height>1296</height>
</resolution>
<language>English</language>
<sdCard>1</sdCard>
<ptzMode>pt</ptzMode>
<typeInfo>IPC</typeInfo>
<softVer>50331763</softVer>
<hardVer>0</hardVer>
<panelVer>0</panelVer>
<hdChannel1>0</hdChannel1>
<hdChannel2>0</hdChannel2>
<hdChannel3>0</hdChannel3>
<hdChannel4>0</hdChannel4>
<norm>NTSC</norm>
<osdFormat>MDY</osdFormat>
<B485>0</B485>
<supportAutoUpdate>0</supportAutoUpdate>
<userVer>1</userVer>
<FrameworkVer>1</FrameworkVer>
<secretCode>00000000841461a1</secretCode>
<authMode>0</authMode>
</DeviceInfo>
<StreamInfoList version="1.1">
<StreamInfo>
<channelBits>1</channelBits>
<encodeTable>
<type>mainStream</type>
<resolution>
<width>2304</width>
<height>1296</height>
</resolution>
<defaultFramerate>15</defaultFramerate>
<defaultBitrate>2560</defaultBitrate>
<framerateTable>15,12,10,8,6,4,2</framerateTable>
<bitrateTable>1024,1536,2048,2560,3072</bitrateTable>
</encodeTable>
<encodeTable>
<type>subStream</type>
<resolution>
<width>704</width>
<height>480</height>
</resolution>
<defaultFramerate>10</defaultFramerate>
<defaultBitrate>384</defaultBitrate>
<framerateTable>15,12,10,8,6,4</framerateTable>
<bitrateTable>128,192,256,384,512,768</bitrateTable>
</encodeTable>
</StreamInfo>
</StreamInfoList>
</body>
```

The new xml keys in there are:

```xml
<FrameworkVer>1</FrameworkVer>
<secretCode>00000000841461a1</secretCode>
<authMode>0</authMode>
```

The `secretCode` in there sounds interesting

p.s. Full diff of the xml is here

```diff
--- before.xml	2021-04-15 09:14:40.000000000 +0700
+++ after.xml	2021-04-15 09:14:53.000000000 +0700
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8" ?>
 <body>
 <DeviceInfo version="1.1">
-<firmVersion>00000000000000</firmVersion>
+<firmVersion>00000000065536</firmVersion>
 <IOInputPortNum>0</IOInputPortNum>
 <IOOutputPortNum>0</IOOutputPortNum>
 <diskNum>0</diskNum>
@@ -19,7 +19,7 @@
 <sdCard>1</sdCard>
 <ptzMode>pt</ptzMode>
 <typeInfo>IPC</typeInfo>
-<softVer>33555019</softVer>
+<softVer>50331763</softVer>
 <hardVer>0</hardVer>
 <panelVer>0</panelVer>
 <hdChannel1>0</hdChannel1>
@@ -27,10 +27,13 @@
 <hdChannel3>0</hdChannel3>
 <hdChannel4>0</hdChannel4>
 <norm>NTSC</norm>
-<osdFormat>DMY</osdFormat>
+<osdFormat>MDY</osdFormat>
 <B485>0</B485>
 <supportAutoUpdate>0</supportAutoUpdate>
 <userVer>1</userVer>
+<FrameworkVer>1</FrameworkVer>
+<secretCode>00000000841461a1</secretCode>
+<authMode>0</authMode>
 </DeviceInfo>
 <StreamInfoList version="1.1">
 <StreamInfo>
@@ -49,13 +52,13 @@
 <encodeTable>
 <type>subStream</type>
 <resolution>
-<width>896</width>
-<height>512</height>
+<width>704</width>
+<height>480</height>
 </resolution>
-<defaultFramerate>15</defaultFramerate>
-<defaultBitrate>512</defaultBitrate>
-<framerateTable>15,12,10,8,6,4,2</framerateTable>
-<bitrateTable>128,256,384,512,768,1024</bitrateTable>
+<defaultFramerate>10</defaultFramerate>
+<defaultBitrate>384</defaultBitrate>
+<framerateTable>15,12,10,8,6,4</framerateTable>
+<bitrateTable>128,192,256,384,512,768</bitrateTable>
 </encodeTable>
 </StreamInfo>
 </StreamInfoList>
```

**Reply by QuantumEntangledAndy** [reply]
After this messages appear the same in the header except that the xml is not decoding

**Reply by thirtythreeforty** [reply]
Ok great work y'all. So from decompiling the firmware, `0x2` is the new mode (there's an `if(1) .. else if(2)` block that does the Charlie routine or AES).

Then, I learned the AES mode is CFB 128 ([this mbedtls function](https://tls.mbed.org/api/aes_8h.html#a944946adabbc344f2c6cf6e6f51a21e3)), and the IV which is a constant string `0123456789abcdef` (doh!). I believe the IV is reset for every message (doh!). The key is 128 bit...

>The new xml keys in there are:
>```
><FrameworkVer>1</FrameworkVer>
><secretCode>00000000841461a1</secretCode>
><authMode>0</authMode>
>```

oh look, the `secretCode` is 16 bytes long.  I think that likely wraps it up.

**Reply by QuantumEntangledAndy** [reply]
I have been playing aroung in https://play.rust-lang.org/ with this code

```rust
use log::*;
use env_logger::Env;
use openssl::symm::*;

const KEY: &'static [u8] = b"00000000841461a1";
const IV: &'static [u8] = b"0123456789abcdef";
const CIPHERTEXT: &'static str = "pi+Bw1QZLWgn1KqParxvviM3lr7D26MDjriD+L0//LGmrBMPTEgNrCCJfNKxA0MEuYGVGX0o4cL33XK8v6Nn0RF8MJub8M3A1dGSkPvYTGYZwA8OLQ==";
const HEADER_SIZE: usize = 0;

fn main() {
    env_logger::Builder::from_env(Env::default().default_filter_or("info")).init();
    
    let decoded = base64::decode(&CIPHERTEXT).unwrap();
    let ciphertext = &decoded[HEADER_SIZE..];
    
    info!("KEY: {:x?}, len: {}", KEY, KEY.len());
    info!(" IV: {:x?}, len: {}", IV, IV.len());

    let t = Cipher::aes_128_cfb128();
    let result = decrypt(t, KEY, Some(IV), &ciphertext).unwrap();
    
    info!("Bytes: {:?}", result);
    info!("UTF8: {:?}", unsafe{std::str::from_utf8_unchecked(&result)});
}
```

It is still not decoding.

I had to change the IV to `0123456789abcdef` because it was one character short was that a typo?

**Reply by thirtythreeforty** [reply]
Ah yes, typo. Good catch. I'll play around with it. Maybe there's some initialization that I missed.

**Reply by QuantumEntangledAndy** [reply]
Aes is a block cipher which means the data should be a multiple of the blocksize which should be 16 bytes. My data is 85 bytes which is not a multiple of 16. Is there some sort of header of something here? Or am I misunderstanding aes.

**Reply by thirtythreeforty** [reply]
I don't think you're doing anything wrong... however @DenisMartin0's dump that he sent me, does _not_ contain the `secretCode` in the DeviceInfo struct. So it's possible we're barking up the wrong tree there, although I don't see how else they exchange it, unless it's calculated on the client/server separately. (Strictly speaking, it's equally insecure either way :smiley: )

**Reply by thirtythreeforty** [reply]
Yep, based on more strings in the Android app, this is a secretCode for push messaging. So, I'll keep looking for how the key is derived.

**Reply by DenisMartin0** [reply]
AES-128 is confirmed, CFB is confirmed, IV is confirmed
The key changes at every login. It's part of a longer 32 bytes string "31 chars + null".
The key only uses 16 first bytes of it.
The longer string looks like a MD5 hash of something, with the last byte missing (Similar to username & pwd shorten md5)
Seems like this string is only based on CAPS & num (0..9A..F)

![image](https://user-images.githubusercontent.com/36104617/114907084-ff083b80-9e1a-11eb-93b7-043bc8d7076a.png)

**Reply by thirtythreeforty** [reply]
Nice. What's your debug workflow for intercepting the key? I'm currently trying to set up WinDbg + Ghidra.

Based on more static analysis the MD5sum is computed with the same janky setup that Neolink already implements (bottom of bc_protocol.rs). The input is two strings joined by a dash as in `{}-{}` - but I don't yet know which strings. I blindly tried the hashed username and password but that didn't work (although I may have been holding it wrong).

**Reply by DenisMartin0** [reply]
IDA, but not easy because there are three different processes... I try to attach to each one until I find the dll loaded as a module

**Reply by thirtythreeforty** [reply]
Ah that's rich folk software 😛 

Can you break at the call to `Md5_string_encrypt` and grab the argument? Don't know if IDA infers types but the first argument is `const char*` I think:

```
uint32_t Md5_string_encrypt(char *src, char *dest);
```

If I can see an example I think we can work out what the semantics are.

**Reply by DenisMartin0** [reply]
It's IDA freeware 😋, feel free to download it.
I''ll try that in a few hours and let you know

**Reply by DenisMartin0** [reply]
@thirtythreeforty There is no such function in the dll, or it's not a 'named' one. Any clue on its location in the assembly ? or any hex sequence/fingerprint for it ? for me to search for it.

**Reply by thirtythreeforty** [reply]
I will look and see

**Reply by twistedddx** [reply]
> Seems like this string is only based on CAPS & num (0..9A..F)

Maybe MAC address?

**Reply by QuantumEntangledAndy** [reply]
I've been doing a ghidra of the camera programs stored in the firmware update. The key is stored in the `c_client_session` structure at offset `0x7c`. I did a search for all things that reference the `c_client_session` then filtered by `0x7c`. This member of the sturucture at `0x7c` is read in many places and is always in the context of the AES key. Only in one place is this member written to. It's in a function that seems to deal with logins and password checking. I will keep digging into it.

**Reply by QuantumEntangledAndy** [reply]
The function also deals with message length checks and status codes. So I think it's parsing the BC header too but not sure.

**Reply by QuantumEntangledAndy** [reply]
Ok so I also had a look into the mac binaries using Ghidra incase that was any easier. This is the block that sets the key

```c
ulong UndefinedFunction_003c63c7
                (char *param_1,size_t param_2,undefined8 param_3,undefined8 param_4,char *param_5)

{
  uint uVar1;
  long *plVar2;
  long lVar3;
  int iVar4;
  undefined4 uVar5;
  char *pcVar6;
  ulong uVar7;
  uint uVar8;
  long unaff_RBP;
  
  iVar4 = __stubs::_snprintf(param_1,param_2,param_5);
  *(int *)(unaff_RBP + -0xbd54) = iVar4;
  _Md5_string_encrypt((uchar *)(unaff_RBP + -0xa310),(undefined8 *)(unaff_RBP + -0xa210));
  if (*(long *)(*(long *)(unaff_RBP + -0xbc80) + 0x7068) != 0) {
    NetMsgClient::setXmlAesEncryptKey
              (*(NetMsgClient **)(*(long *)(unaff_RBP + -0xbc80) + 0x7068),
               (char *)(unaff_RBP + -0xa210));

```

Unfortuantly it looks pretty complex here and I think it might be using a global variable because of the `unaff_RB` not sure where that register is assigned. Really would be nice to have a breakpoint on that `_Md5_string_encrypt`. Will probably be my next job when I get time.

It's getting late here hope you have more luck than me.

**Reply by DenisMartin0** [reply]
@thirtythreeforty @twistedddx @QuantumEntangledAndy Done ;)
If you wanna try:
sprintf(key, "%s-%s", nonce, password);
AES_USER_KEY = LEFT(UPPERCASE(MD5(key)), 16); // The MD5 is the one with the strange missing last byte which is replaced by a '\0'; but no matter as we'll keep only the 16 left chars

**Reply by DenisMartin0** [reply]
As a bonus, here is the nonce partial reverse:
"`6079eb13-iaoVvfQIhHqhPSAsudEc`"
`6079eb13 `-> unix timestamp in hex
`iaoVvfQIhH `-> simple transform from unix timestamp, and/or related to datetime (ie: from the same timestamp, you'll get the same data); looks like a base64
`qhPSAsudEc `-> Sometime it's constant between two runs of the client... looks like a base64 too

**Reply by thirtythreeforty** [reply]
Awesome work thank you! Does IDA connect to a debugger?

That's easy enough to implement. I'll see if I can work up a patch tonight.

Now, separately the Lua library needs to be able to decode this. I wonder if Wireshark already provides decryption primitives.

**Reply by thirtythreeforty** [reply]
For the nonce, the NVR seems to create different nonces than the cameras. I wonder why.

**Reply by QuantumEntangledAndy** [reply]
For reference here is a working decoder:

```rust
use log::*;
use env_logger::Env;
use openssl::symm::*;
use md5::{Md5, Digest};

const NONCE: &'static str = "EFC616738D65B1D0";
const PWD: &'static str = "testPass";
const IV: &'static [u8] = b"0123456789abcdef";
const CIPHERTEXT: &'static str = "x2n/hpPivHqK8s3DIv+5kF2mj/+krYGpvTkSmspBfbdQRikDABCi+2sYSRnVJM7bxi6jquqRR2hNPhscATMtaB4KozAWAB4tqmAan/2/LnR7uiBhOdx1ScrMZPfi7vlzXZnWuPU3Za+gR8liJVVRc8JQ6PlROHBRjyQ6i+HN+FLFWCchai9edOxllyHxznlfWUVP1iKrSx6U26U7zycI/wXgyyxXGpsk/v90uWfYVmSCdHJNPKSrLZnbutYCOPmNnjc4X2gAZJRZLYhPW/Wv+GS3eeV2C/KbcWXFXWUUVYGjuh01QPUGuOZHm5Jljhu/bX9M5gYdKc1H34XxY6cILOsTOvi481Zayo5k4euI8fsWH0czDGytmtYDRN06Z9visb28dXzYAjJkPicSCJjNp/hp7Zg1uyN/McLuvx/kljUIvP24l4/4XT+kRR98H8ipdirP+GiE49g0Ko73xkajQvvb1nMsOhxLpw1MQOq3K5AsQaY1T0Fodf2Eves2o2LW6GlvSuM3vf7DK9JRWI3bindjDKhwK1Pn2tRK4uUZ0vybm6aMAE1fufhLy0054+E5XSKDnIBJp+of6s9zBcgCUClznRaNuPZ1muGMIi5j3bldU7iDQS45f8SxdDu0EcBXBZ17n8lPFSn0K4DSwHxauAUS";
const HEADER_SIZE: usize = 0;
const TAIL_SIZE: usize = 0;

fn main() {
    env_logger::Builder::from_env(Env::default().default_filter_or("info")).init();
    
    let decoded = base64::decode(&CIPHERTEXT).unwrap();
    let ciphertext = &decoded[HEADER_SIZE..(decoded.len()-TAIL_SIZE)];
    
    let key_phrase = format!("{}-{}", NONCE, PWD);
    let key_phrase_hash = md5_string(&key_phrase, Md5Trunc::ZeroLast).to_uppercase().into_bytes();
    
    let key = &key_phrase_hash[0..16];
    let iv = &IV[0..16];
    
    info!("KEY: {:x?}, len: {}", key, key.len());
    info!(" IV: {:x?}, len: {}", iv, iv.len());
    info!("ciphertext: {:x?}, len: {}", ciphertext, ciphertext.len());

    let t = Cipher::aes_128_cfb128();
    let result = decrypt(t, key, Some(iv), &ciphertext).unwrap();
    
    info!("Bytes: {:x?}, len: {}", result, result.len());
    info!("UTF8: {:?}", unsafe{std::str::from_utf8_unchecked(&result)});
}
#[derive(PartialEq, Eq)]
enum Md5Trunc {
    ZeroLast,
    Truncate,
}

fn md5_string(input: &str, trunc: Md5Trunc) -> String {
    let mut md5 = format!("{:X}\0", Md5::digest(input.as_bytes()));
    md5.replace_range(31.., if trunc == Md5Trunc::Truncate { "" } else { "\0" });
    md5
}
```

**Reply by QuantumEntangledAndy** [reply]
Ok I have a working PR here #83. Its an update of my Protocol update. So it includes the new BC models that we worked out earlier and it knows about message ids and status codes. Please let me know what you think

**Reply by QuantumEntangledAndy** [reply]
If we merge this PR I hope to get HD and SD streams from a single login using message ids

**Reply by thirtythreeforty** [reply]
Nice you beat me to it. I'll comment there.

**Reply by QuantumEntangledAndy** [reply]
I had 90% done already because I was working on an unencrypted protocol update (@twistedddx  requested it ages ago for his camera) which needed a switch in the same places this needed a switch.
